crates/rfc5321/src/tls.rs

```diff
@@ -1,11 +1,72 @@
 #![cfg(feature = "client")]
 use hickory_proto::rr::rdata::TLSA;
+use lruttl::LruCacheWithTtl;
 use openssl::ssl::SslOptions;
-use std::sync::Arc;
+use std::sync::{Arc, LazyLock, Mutex};
 use tokio_rustls::rustls::crypto::{aws_lc_rs as provider, CryptoProvider};
 use tokio_rustls::rustls::{ClientConfig, SupportedCipherSuite};
 use tokio_rustls::TlsConnector;
 
+struct RustlsCacheKey {
+    insecure: bool,
+    rustls_cipher_suites: Vec<SupportedCipherSuite>,
+}
+
+// SupportedCipherSuite has a PartialEq impl but not an Eq impl.
+// Since we need RustlsCacheKey to be Hash we cannot simply derive
+// PartialEq and then add an explicit impl for Eq on RustlsCacheKey
+// because we don't know the implementation details of the underlying
+// PartialEq impl. So we define our own here where we explicitly compare
+// the suite names. This may not be strictly necessary, but it seems
+// wise to be robust to possible future weirdness in that type, and
+// to be certain that our Hash impl is consistent with the Eq impl.
+impl std::cmp::PartialEq for RustlsCacheKey {
+    fn eq(&self, other: &RustlsCacheKey) -> bool {
+        if self.insecure != other.insecure {
+            return false;
+        }
+        self.rustls_cipher_suites
+            .iter()
+            .map(|s| s.suite())
+            .eq(other.rustls_cipher_suites.iter().map(|s| s.suite()))
+    }
+}
+
+impl std::cmp::Eq for RustlsCacheKey {}
+
+impl std::hash::Hash for RustlsCacheKey {
+    fn hash<H>(&self, hasher: &mut H)
+    where
+        H: std::hash::Hasher,
+    {
+        self.insecure.hash(hasher);
+        for suite in &self.rustls_cipher_suites {
+            suite.suite().as_str().hash(hasher);
+        }
+    }
+}
+
+static RUSTLS_CACHE: LazyLock<Mutex<LruCacheWithTtl<RustlsCacheKey, Arc<ClientConfig>>>> =
+    LazyLock::new(|| Mutex::new(LruCacheWithTtl::new(32)));
+
+impl RustlsCacheKey {
+    fn get(&self) -> Option<Arc<ClientConfig>> {
+        RUSTLS_CACHE.lock().unwrap().get(self)
```

Comment:
Maybe consider using the parking_lot primitives? In addition to being faster they don't have the poisoning issue so you don't have to unwrap all over the place.